Implement pluggable Lineage in Java SDK

shnapz opened the pull request:
Addresses #36790: "[Feature Request]: Make lineage tracking pluggable"

## Changes

- **Created** `org.apache.beam.sdk.lineage.LineageBase` - Plugin interface with single `add()` method
- **Created** `org.apache.beam.sdk.lineage.LineageOptions` - Pipeline option (`--lineageType`) for explicit plugin selection
- **Refactored** `org.apache.beam.sdk.metrics.Lineage` - Hardcoded metrics → Delegation to `LineageBase` plugins
- **Created** `org.apache.beam.sdk.metrics.BoundedTrieMetricsLineage` - Default BoundedTrie-based implementation
- **Created** `org.apache.beam.sdk.metrics.StringSetMetricsLineage` - Default StringSet-based implementation
- **Added** Plugin initialization in `FileSystems.setDefaultPipelineOptions()`

## Architecture

**Before (master)**: `Lineage` was a concrete class hardcoded to use Beam metrics:

```java
public class Lineage {
  private static final Lineage SOURCES = new Lineage(Type.SOURCE);
  private static final Lineage SINKS = new Lineage(Type.SINK);
  private final Metric metric;  // Hardcoded to Beam metrics

  private Lineage(Type type) {
    this.metric = Metrics.stringSet(LINEAGE_NAMESPACE, type.toString());
  }

  public void add(Iterable<String> segments) {
    ((StringSet) metric).add(String.join("", segments));  // Always metrics
  }
}
```

**After (this PR)**: Clean separation via composition pattern:

```java
// Plugin contract (simple interface)
public interface LineageBase {
  void add(Iterable<String> rollupSegments);
}

// Public API (final facade delegating to plugin)
public final class Lineage {
  private final LineageBase delegate;  // Plugin implementation

  private Lineage(LineageBase delegate) {
    this.delegate = delegate;
  }

  // Delegates to plugin
  public void add(Iterable<String> segments) {
    delegate.add(segments);
  }

  // Convenience overloads
  public void add(String system, Iterable<String> segments) { ... }
  public void add(String system, String subtype, ...) { ... }

  // Static utilities (unchanged)
  public static Lineage getSources() { ... }
  public static Lineage getSinks() { ... }
  public static String wrapSegment(String value) { ... }
  public static Set<String> query(MetricResults results, Type type) { ... }
}

// Default implementations (type-safe, no casts)
class BoundedTrieMetricsLineage implements LineageBase {
  private final BoundedTrie metric;
  @Override
  public void add(Iterable<String> segments) {
    metric.add(ImmutableList.copyOf(segments));
  }
}

class StringSetMetricsLineage implements LineageBase {
  private final StringSet metric;
  @Override
  public void add(Iterable<String> segments) {
    metric.add(String.join("", segments));
  }
}
```

**Plugin Selection**: Explicit via `--lineageType` pipeline option. The specified class must implement `LineageBase` and have a public `(PipelineOptions, Lineage.LineageDirection)` constructor. When not set, defaults to metrics-based lineage.

**Backward Compatibility**: ✅ All existing code works unchanged (24+ call sites, static utilities, enums).

## Why Pluggable Lineage?

### 1. Runner Fragmentation

Metrics-based lineage is **scattered across runners** with inconsistent support:
- **Dataflow**: Real-time metrics export to Cloud Monitoring
- **Flink**: Batch-only aggregation (no streaming support yet)
- **Spark/Direct**: Varying levels of support

**Impact**: Multi-runner organizations must consolidate lineage from different metrics backends, each with different APIs and formats.

**Plugin Solution**: Single implementation works consistently across all runners.

### 2. Enterprise Integration

Organizations with existing lineage infrastructure need:
- **Direct API integration** (Atlan, Collibra, Marquez, DataHub, OpenLineage)
- **Custom metadata enrichment** not in metrics subsystem

**Example**: Flyte workflow executing a Beam pipeline needs to tag lineage with Flyte execution ID and cost allocation. This context exists in the orchestrator, not in Beam workers' metrics.

### 3. Standard Formats

OpenLineage is the industry standard. Plugin enables direct emission vs. export metrics → parse → transform → send.

## Initialization

`Lineage.setDefaultPipelineOptions(options)` is called from `FileSystems.setDefaultPipelineOptions()` (same pattern as `Metrics`).

**Rationale**: `FileSystems.setDefaultPipelineOptions()` is called at 48+ locations covering all execution scenarios (pipeline construction, worker startup, deserialization).

**Known Limitation**: Follows existing `FileSystems` pattern despite known issues ([#18430](https://github.com/apache/beam/issues/18430)). Architectural improvements would address all subsystems together.

## Thread Safety

Uses synchronized block for thread-safe initialization. `setDefaultPipelineOptions` always re-initializes under a lock, ensuring plugins pick up any options changes.

## Example: OpenLineage Plugin

_For demonstration only (OpenLineage integration out of scope)_

```java
// 1. Implement LineageBase with required constructor
class OpenLineageReporter implements LineageBase {
  private final String endpoint;
  private final Lineage.LineageDirection direction;

  // Required constructor for --lineageType instantiation
  public OpenLineageReporter(PipelineOptions options, Lineage.LineageDirection direction) {
    OpenLineageOptions opts = options.as(OpenLineageOptions.class);
    this.endpoint = opts.getOpenLineageUrl();
    this.direction = direction;
  }

  @Override
  public void add(Iterable<String> rollupSegments) {
    String fqn = String.join("", rollupSegments);
    // POST to OpenLineage API with workflow context
    sendToOpenLineage(endpoint, direction, fqn);
  }
}

// 2. Plugin options
public interface OpenLineageOptions extends PipelineOptions {
  @Description("OpenLineage endpoint URL")
  String getOpenLineageUrl();
  void setOpenLineageUrl(String url);
}

// 3. Usage — select plugin via pipeline option
PipelineOptions options = PipelineOptionsFactory.create();
options.as(OpenLineageOptions.class).setOpenLineageUrl("https://lineage-api.example.com");
options.as(LineageOptions.class).setLineageType(OpenLineageReporter.class);
Pipeline p = Pipeline.create(options);

// Or from command line:
// --lineageType=com.example.OpenLineageReporter --openLineageUrl=https://lineage-api.example.com
```

shnapz committed:
LineageReporter plugin initial implementation

shnapz committed:
Change to Lineage as a base abstract class

shnapz committed:
Add tests

shnapz committed:
Add a TestPipeline integration test

shnapz committed:
Initialize Lineage from FileSystem

shnapz committed:
Fix formatting

shnapz committed:
Fix Flaky JmsIOTest

shnapz committed:
Adding to change log and fixing style error

shnapz committed:
fix build

shnapz committed:
fix tests

shnapz committed:
Improve test logging

shnapz committed:
fix test

shnapz committed:
fix formatting

shnapz committed:
Revert unnecessary fixes in tests

kennknowles reviewed (changes requested):
Adding @rohitsinha54 who knows more about lineage. My comments are just on repo code style. I'd like to know more about the alternative implementation(s). It makes sense that you might want to skip going through metrics and directly report lineage elsewhere. There should probably be a quick sketch of a design for whether there is just one plugin active, or if all the lineage observers receive all events, etc.

shnapz committed:
Extract separate LineageBase interface

shnapz committed:
small fixes

Abacn commented on sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/MetricsLineage.java:
```diff
@@ -0,0 +1,60 @@
+  public void add(final Iterable<String> rollupSegments) {
+    ImmutableList<String> segments = ImmutableList.copyOf(rollupSegments);
+    if (MetricsFlag.lineageRollupEnabled()) {
+      ((BoundedTrie) this.metric).add(segments);
```
This adds an assumption (and coupling) between lineageRollupEnabled<->BoundedTrie. With common interface we can now avoid unchecked cast by implementing CounterMetricsLineage and TrieMetricsLineage separately.

shnapz committed:
Address PR review

shnapz committed:
Added LineageOptions registration

shnapz committed:
remove unused files

shnapz committed:
fix failing test with moquito

shnapz committed:
Address review comments

shnapz committed:
Simplify Lineage thread sync during init

shnapz committed:
target 2.73.0

shnapz commented:
hey @Abacn did you have a change to review latest changes? thank you

Abacn commented:
Thanks for persistence. There were lots of back-and-forth on setDefaultPipelineOptions lately. The latest commit still runs SOURCES.set, SINKS.set every time gets called. Let me summarizes requirements

- When LineageType option isn't set, there shouldn't be overriden every time setDefaultPipelineOptions invoked

- When LineageType option has been set before, we check necessity of such overwrite.

I think the version before the latest one could be a good start point. Instead of computeSet, we can do something like

```
Class<? extends LineageBase> lineageClass = options.as(LineageOptions.class).getLineageType();
@Nullable LineageBase currentLineage = SOURCES.get();
if (SOURCES.get() == null || (lineageClass != null && !lineageClass.isInstance(currentLineage)) {
  // we need to re-generate
  synchronized {
    ...
  }
}
```

shnapz committed:
improve setDefaultPipelineOptions concurrency

shnapz commented:
@Abacn thanks, pushed a rewrite along these lines:                                                                                                                    
                                                                                                                                                                 
- Double-checked lock with canSkipInit(requestedType): skip when `SOURCES != null && (requestedType == null || requestedType.equals(CURRENT_LINEAGE_TYPE))`.
- Tracking the active type in a separate volatile Class<? extends LineageBase> CURRENT_LINEAGE_TYPE. Functionally equivalent to your sketch.
- Switched SOURCES/SINKS to volatile.
- Also dropped the lazy fallback in getSources()/getSinks() - now checkNotNull with a clear message, so we never silently swap plugin lineage for defaults. WDYT?
                                                                                              
PTAL

FYI Two tests in StreamingDataflowWorkerTest failed, flaky and unrelated

shnapz committed:
fix spotless

Abacn commented on CHANGES.md:
```diff
@@ -106,6 +106,7 @@
 
 ## New Features / Improvements
 
+* Added plugin mechanism to support different Lineage implementations (Java) ([#36790](https://github.com/apache/beam/issues/36790)).
```
This goes to Beam 2.74.0 for now.

shnapz replied:
Done, moved to 2.74 section

Abacn commented on sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/BoundedTrieMetricsLineage.java:
```diff
@@ -0,0 +1,44 @@
+
+  private final BoundedTrie metric;
+
+  BoundedTrieMetricsLineage(Lineage.LineageDirection direction) {
```
Both StringSetMetricsLineage and StringSetMetricsLineage's constructor doesn't satisfy its own LineageBase interface documentation:

```
<p>Implementations must provide a public constructor accepting ({@link
 * org.apache.beam.sdk.options.PipelineOptions}, {@link
 * org.apache.beam.sdk.metrics.Lineage.LineageDirection}).
```

please fix

shnapz replied:
Fixed - added this parameter to build-in plugins as well, now all are aligned with LineageBase

Abacn reviewed (commented):
Thanks, I see the multiple intialization is resolved.

shnapz committed:
final fixes

Abacn commented on sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/BoundedTrieMetricsLineage.java:
```diff
@@ -29,7 +30,8 @@ class BoundedTrieMetricsLineage implements LineageBase {
 
-  BoundedTrieMetricsLineage(Lineage.LineageDirection direction) {
+  @SuppressWarnings("unused")
+  public BoundedTrieMetricsLineage(PipelineOptions options, Lineage.LineageDirection direction) {
```
nit: we can just name it as `PipelineOptions options`

shnapz replied:
Addressed this

Abacn replied:
sorry, I meant `PipelineOptions unused` can also make check happy (`@SuppressWarnings("unused")`). This isn't a required change.

shnapz replied:
oh sorry, I misread your earlier nit as a rename suggestion
btw: Java unit tests passed green and Error-Prone UnusedVariable would fail but it's not flagging the param here (likely because public constructors get a pass).

Happy to put the annotation back if you prefer

Abacn replied:
Yeah, either need `suppressWarning` or name the unused parameter as "unused" could make errorprone happy

shnapz replied:
okay, I reverted back to `@SuppressWarnings("unused")`

Abacn reviewed (approved):
Thanks! Will merge once tests passed

shnapz committed:
address review comments

shnapz committed:
put suppressions back

shnapz commented:
@Abacn all checks are green 👍

Abacn commented on sdks/java/core/src/test/java/org/apache/beam/sdk/lineage/LineagePluginTest.java:
```diff
@@ -0,0 +1,298 @@
+  }
+
+  /** Helper to create a TestPipeline with test lineage configured. */
+  private TestPipeline createTestPipelineWithLineage() {
```
Here it creates TestPipeline in an unconventional way that ignores beamTestPipelineOptions, breaking validate runners

At this stage I would suggest we revert this change at the moment, since it breaks multiple tests+misconfigured tests, then split this PR into smaller ones

Abacn replied:
Fully understand this is a difficult task as it involves refactoring and decoupling core component thus very easy to break. Thanks again for your persistence

Abacn replied:
alternatively, a forward fix: #38296

shnapz replied:
@Abacn your fix looks great! I hope it gets merged!

shnapz commented:
@Abacn now #38296 is merged, is there any issue left?
